azure-kusto-data/azure/kusto/data/_token_providers.py

```diff
@@ -0,0 +1,426 @@
+    AZ_REFRESH_TOKEN = "refreshToken"
+    AZ_USER_ID = "userId"
+    AZ_AUTHORITY = "_authority"
+    AZ_CLIENT_ID = "_clientId"
+
+
+class TokenProviderBase(abc.ABC):
+    """
+    This base class abstracts token acquisition for all implementations.
+    The class is build for Lazy initialization, so that the first call, take on instantiation of 'heavy' long-lived class members
+    """
+
+    _initialized = False
+    _kusto_uri = None
+    _cloud_info = None
+    _scopes = None
+
+    def __init__(self, kusto_uri: str):
+        self._kusto_uri = kusto_uri
+        if kusto_uri is not None:
+            self._scopes = [kusto_uri + ".defult" if kusto_uri.endswith("/") else kusto_uri + "/.default"]
+            self._cloud_info = CloudSettings.get_cloud_info(kusto_uri)
+
+    def get_token(self):
+        """ Get a token silently from cache or authenticate if cached token is not found """
+        if not self._initialized:
+            self._init_impl()
+            self._initialized = True
+
+        token = self._get_token_silent_impl()
+        if token is None:
+            token = self._get_token_impl()
+
+        return self._valid_token_or_throw(token)
+
+    @staticmethod
+    @abc.abstractmethod
+    def name() -> str:
+        """ return the provider class name """
+        pass
+
+    @abc.abstractmethod
+    def context(self) -> dict:
+        """ return a secret-free context for error reporting """
+        pass
+
+    @abc.abstractmethod
+    def _init_impl(self):
+        """ Implement any "heavy" first time initializations here """
+        pass
+
+    @abc.abstractmethod
+    def _get_token_impl(self) -> dict:
+        """ implement actual token acquisition here """
+        pass
+
+    @abc.abstractmethod
+    def _get_token_silent_impl(self) -> dict:
+        """ Implement cache checks here, return None if cache check fails """
+        pass
+
+    @staticmethod
+    def _valid_token_or_none(token: dict) -> dict:
+        if token is None or TokenConstants.MSAL_ERROR in token:
+            return None
+
+        return token
+
+    def _valid_token_or_throw(self, token: dict) -> dict:
+        if token is None:
+            raise KustoClientError(self.name() + " failed to obtain a token")
+
+        if TokenConstants.MSAL_ERROR in token:
+            message = self.name() + " failed to obtain a token: " + token[TokenConstants.MSAL_ERROR]
+            if TokenConstants.MSAL_ERROR_DESCRIPTION in token:
+                message = message + "\n" + token[TokenConstants.MSAL_ERROR_DESCRIPTION]
+
+            raise KustoClientError(message)
+
+        return token
+
+
+class BasicTokenProvider(TokenProviderBase):
+    """ Basic Token Provider keeps and returns a token received on construction """
+
+    def __init__(self, token: str):
+        super().__init__(None)
+        self._token = token
+
+    @staticmethod
+    def name() -> str:
+        return "BasicTokenProvider"
+
+    def context(self) -> dict:

```

Comment:
Can't we make those methods (context, init_impl, _get_token_impl) virtual instead of abstract ?

Reply:
They don't have an implementation to begin with, so I think abstract is more fitting.
Why use virtual?

Reply:
I see in various child classes:
`    def _init_impl(self):`
`        pass`

 `   def _get_token_impl(self) -> dict:`
 `       return None`

Reply:
In some implementations these operations are meaningless, but that's for the implementer to decide.
Not sure how well does ABC enforce abstraction not being implemented, but I would prefer any new implementation to consider what to do in each function and not rely on the base class.